iOS/Flows/Conversation/Cells/TimeMachineCollectionViewLayout.swift

```diff
@@ -0,0 +1,304 @@
+        }
+    }
+
+    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
+        // The positions of the items need to be recalculated for every change to the bounds.
+        return true
+    }
+
+    override func invalidateLayout() {
+        super.invalidateLayout()
+
+        // Clear the layout attributes caches.
+        self.cellLayoutAttributes.removeAll()
+        self.zRangesDict.removeAll()
+    }
+
+    override func prepare() {
+        self.prepareZRanges()
+
+        // Calculate and cache the layout attributes for the items in each section.
+        self.forEachIndexPath { indexPath in
+            self.cellLayoutAttributes[indexPath] = self.layoutAttributesForItem(at: indexPath)
+        }
+    }
+
+    /// Updates the z ranges dictionary for all items.
+    private func prepareZRanges() {
+        guard let dataSource = self.dataSource else { return }
+
+        // Get all of the items and sort them by value. This combines all the sections into a flat list.
+        var sortedItemIndexPaths: [IndexPath] = []
+        self.forEachIndexPath { indexPath in
+            sortedItemIndexPaths.append(indexPath)
+        }
+        sortedItemIndexPaths.sort { indexPath1, indexPath2 in
+            let message1 = dataSource.getMessage(at: indexPath1)!
+            let message2 = dataSource.getMessage(at: indexPath2)!
+            return message1.createdAt < message2.createdAt
+        }
+
+        for (sortedItemsIndex, indexPath) in sortedItemIndexPaths.enumerated() {
+            let currentSection = indexPath.section
+            let currentItemIndex = indexPath.item
+
+            var startZ: CGFloat = 0
+            // Each item's z range starts after the end of the previous item's range within its section.
+            if let previousRangeInSection = self.zRangesDict[IndexPath(item: currentItemIndex - 1,
+                                                                      section: currentSection)] {
+
+                startZ = previousRangeInSection.upperBound + self.itemSize.height
+            }
+
+            var endZ = startZ
+            // Each item's z range ends before the beginning of the next item's range from within its section.
+            if sortedItemsIndex + 1 < sortedItemIndexPaths.count {
+                for nextIndex in (sortedItemsIndex+1)..<sortedItemIndexPaths.count {
+                    let nextIndexPath = sortedItemIndexPaths[nextIndex]
+
+                    if currentSection == nextIndexPath.section {
+                        endZ = CGFloat(nextIndex) * self.itemSize.height - self.itemSize.height
+                        break
+                    } else if nextIndex == sortedItemIndexPaths.count - 1 {
+                        endZ = CGFloat(nextIndex) * self.itemSize.height
+                        break
+                    }
+                }
+            }
+
+            self.zRangesDict[indexPath] = startZ..<endZ
```

Comment:
Each item has a range of z values where it is the frontmost in the stack. This function is calculating that.